### PR TITLE
Checked if error handler is default before setting headers

### PR DESCRIPTION
**Pull request by SwampApe**
Seems like the fix suggested in #2602 did not change the behavior accurately. I added an extra isDefaultErrorHandler attribute to Context and checked for it before setting headers. 
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: lib/reply.js
Comment (mcollina) [inline lib/reply.js:507]: Can we move this behavior in the default errorHandler instead?

**Review by mcollina — CHANGES_REQUESTED** [review]

File: lib/reply.js
Comment (SwampApe) [reply lib/reply.js:507]: If the error handler sends a new error like the default error handler does, we want to set the headers.

Comment (SwampApe) [reply lib/reply.js:507]: It would be possible to move the behavior to default errorHandler and check reply[kReplyErrorHandlerCalled] === true where it was originally to see if we set the headers. That way Context doesn't have an additional attribute.

Comment (mcollina) [reply lib/reply.js:507]: I would prefer to not add another field to context, go ahead and move it.

File: lib/context.js
Comment (mcollina) [inline lib/context.js:21]: This is not needed anymore, right?

Comment (SwampApe) [reply lib/context.js:21]: That's correct my mind just skipped over that.

**Review by mcollina — COMMENTED** [review]
Could you please update the docs as well? I think this new behavior should be documented.

**Review by mcollina — APPROVED** [review]
lgtm

---

We need to resurrect the next branch, as this is semver-major.

**Comment by SwampApe** [comment]
Sorry I'm new to this whole thing. Is there any further work required from me?

**Review by L2jLiga — APPROVED** [review]
lgtm

**Comment by jsumners** [comment]
@mcollina done. PR has been retargeted.

File: lib/reply.js
Comment (mcollina) [inline lib/reply.js:516]: Can you avoid duplication of this block and the one you added in the default error handler by placing them in a new `.js` file? Thanks.

**Comment by mcollina** [comment]
@SwampApe can you address https://github.com/fastify/fastify/pull/2609#discussion_r504596290? Thanks

**Comment by SwampApe** [comment]
Sorry I'm not sure why but this pull request didn't update when I sent my commit a few hours ago. It should be good now.

File: lib/setErrorHeaders.js
Comment (mcollina) [inline lib/setErrorHeaders.js:5]: Can you please move the extraction of `res` from `reply` inside the function?

```suggestion
function setErrorHeaders (error, reply) {
  const res = reply.raw
```

**Review by mcollina — COMMENTED** [review]
Good work, just a nit.

**Review by mcollina — APPROVED** [review]
lgtm, I think this is great.

Let's wait for a few other to review @delvedor @jsumners @Eomm

File: lib/setErrorHeaders.js
Comment (jsumners) [inline lib/setErrorHeaders.js:10]: How is this treating `undefined` and `null` equivalently?

File: docs/Errors.md
Comment (Eomm) [inline docs/Errors.md:51]: so an error in the error handler, could lead to a 200? - not sure it would be a good thing: in this case the user would miss alert from a monitoring system

In any case, could we add a test for this use case?

File: lib/setErrorHeaders.js
Comment (SwampApe) [reply lib/setErrorHeaders.js:10]: This was here before so I assumed it was true but I just checked and null == undefined -> true.

File: docs/Errors.md
Comment (SwampApe) [reply docs/Errors.md]: An error in the error handler will not lead to a 200. It will follow the default behavior as if a custom error handler isn't specified. I believe there was an existing test case for this but I incorrectly changed the test description earlier but I have just changed it back.

File: test/reply-error.test.js
Comment (RafaelGSS) [inline test/reply-error.test.js:376]: This change is expected? Seems is expected 401 here since the `Error.statusCode = 401`

File: test/reply-error.test.js
Comment (RafaelGSS) [inline test/reply-error.test.js:465]: I think that a check for `Content-Length` here is good as well.

File: test/reply-error.test.js
Comment (SwampApe) [reply test/reply-error.test.js:376]: The new intended behavior is that the statusCode shouldn't be set for a custom error handler. It would only be set if the custom error handler returns a new error or if there is no custom error handler.

**Comment by mcollina** [comment]
I would really love @delvedor and @jsumners opinion before landing this.

**Comment by jsumners** [comment]
It looked to me like @Eomm was still reviewing this. I have been waiting for that process to resolve before I come back.

File: docs/Errors.md
Comment (Eomm) [reply docs/Errors.md]: The sentence is clearer right now 👍 

I searched across the test and I didn't find a test for this use case.
Since there is the new sentence in the readme, could you add the test that verifies that behaviour?

I tried it and I'm not sure it works as expected with

```
fastify.setErrorHandler((err, request, reply) => {
  console.log('error handler')
  throw new Error('ops2')
})

fastify.get('/', async function (request, reply) {
  throw new Error('ops')
})
```

**Review by fox1t — APPROVED** [review]
LGTM!

File: docs/Errors.md
Comment (SwampApe) [reply docs/Errors.md]: I just want to clarify if this test is adequate before I add in a new test.
https://github.com/fastify/fastify/pull/2609/commits/8f9ea6ac91e62d603a180f483f7e21af0374f96d#diff-086d3e4983ab98bad387ce62fb47ff74c0646e94ef5caf4701d096160ea20c06R364-R396

File: test/reply-error.test.js
Comment (SwampApe) [reply test/reply-error.test.js:465]: Added the content-length check

File: docs/Errors.md
Comment (Eomm) [reply docs/Errors.md]: That test checks an error sent, I think "errors thrown by a custom error handler" mean a sync error like my example `throw new Error('ops2')` that is an unexpected behaviour by the developer
I would improve this sentence to avoid confusion like I was 😄

Comment (SwampApe) [reply docs/Errors.md]: It should have the same behavior whether it is thrown or sent because some other part of fastify catches it somewhere. I'm not sure where it is tested but I'll improve the sentence.

File: test/reply-error.test.js
Comment (RafaelGSS) [reply test/reply-error.test.js:376]: So, the test title needs to change, right?

Comment (SwampApe) [reply test/reply-error.test.js:376]: No I don't think so. The error object sent from the custom error handler should still set the status code and header for the response. The new change is that the reply object shouldn't already be touched before going to the custom error handler. I have a new test that checks for that.

File: docs/Errors.md
Comment (SwampApe) [reply docs/Errors.md]: `errors sent or thrown by a custom error handler will be caught safely and routed to Fastify's default error handler for a generic 500 Internal Server Error response.` Is this an adequate description for the behavior?

Comment (SwampApe) [reply docs/Errors.md]: Any thoughts on the current `errors sent or thrown by a custom error handler will be routed to Fastify's default error handler` 
Should I put anything talking about how the reply object will have the default status code / header? I don't think it'd be expected for the reply object to have a status 500 and headers set.

**Review by RafaelGSS — APPROVED** [review]
LGTM.

**Comment by SwampApe** [comment]
It seems like @Eomm is busy, should I keep waiting or is there anything else I need to do @mcollina?

**Comment by Eomm** [comment]
I'm not blocking 👍 

I just remeber I didn't find a test that cover this use case - it should be a lack of the existing tests suite:

```
const fastify = require('fastify')()
fastify.get('/', function (req, reply) {
  const err = new Error('kaboom')
  err.headers = { 'fake-random-header': 'abc' }
  reply.send(err)
})

fastify.setErrorHandler((err, req, res) => {
  err.foo.bar()
})

fastify.inject({
  method: 'GET',
  url: '/'
}, (err, res) => {
  // res must be 500
  console.log({ err, code: res.statusCode, body: res.payload })
})
```

**Comment by SwampApe** [comment]
I've added a new test but I'm not sure why pretty print is failing tests. :/

**Comment by Eomm** [comment]
> I've added a new test for the use case but I'm not sure why pretty print is failing tests. :/

If you merge master, it should be solved 👍 

Ref #2679

**Comment by SwampApe** [comment]
Thanks! Just merged.

**Comment by jsumners** [comment]
Uh oh. That was the wrong move. Now this PR has many commits that do not relate to it. You should `git commit reset <last_sha_in_your_commits>` and then `git rebase master`.

**Review by jsumners — APPROVED** [review]
LGTM

**Comment by SwampApe** [comment]
Okay so I rebased instead. I hope that worked out.

**Comment by jsumners** [comment]
:-/

When I left my LGTM this PR was reduced to the 6 affected files. Since then it has gone back to 81 affected files.

**Comment by SwampApe** [comment]
Should I keep the branch not rebased? The next branch this pr is merging into is many commits behind master.

**Comment by jsumners** [comment]
Oh. Then the rebase should be on the `next` branch, not the `master` branch.

**Comment by jsumners** [comment]
If `next` is out-of-date, then one of the maintainers needs to get that branch updated first.

**Comment by SwampApe** [comment]
In that case it'd be great if next can get updated. The default error handler was moved out of context.js in an earlier commit to master.

**Comment by mcollina** [comment]
I've rebased next on top of master.

**Review by mcollina — APPROVED** [review]
still lgtm

**Comment by SwampApe** [comment]
Sorry for the issues haha. I'm gonna figure out git for a sec.

**Comment by SwampApe** [comment]
Alright I hope this is better 😅

**Review by jsumners — APPROVED** [review]
LGTM

**Comment by SwampApe** [comment]
@Eomm thoughts on the new test?

**Review by delvedor — APPROVED** [review]
LGTM

**Pull request merged**
